src/BenchmarksApps/HttpClientBenchmarks/Clients/HttpClient/Program.cs

```diff
@@ -0,0 +1,254 @@
+            _contentStartTimes = new ConcurrentBag<long>();
+            _contentEndTimes = new ConcurrentBag<long>();
+        }
+
+        using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(duration));
+        var stopwatch = Stopwatch.StartNew();
+
+        var tasks = new List<Task>(_options.NumberOfClients * _options.ConcurrencyPerClient);
+        for (int i = 0; i < _options.NumberOfClients; ++i)
+        {
+            var client = _httpClients[i];
+            for (int j = 0; j < _options.ConcurrencyPerClient; ++j)
+            {
+                switch(_options.Scenario)
+                {
+                    case "get":
+                        tasks.Add(Get(client, cts.Token, collectMetrics));
+                        break;
+                    default:
+                        throw new ArgumentException($"Unknown scenario: {_options.Scenario}");
+                }
+            }
+        }
+
+        await Task.WhenAll(tasks);
+        var elapsed = stopwatch.Elapsed.TotalSeconds;
+
+        if (collectMetrics)
+        {
+            BenchmarksEventSource.Measure("http/successrequests", _successRequests);
+            BenchmarksEventSource.Measure("http/badstatusrequests", _badStatusRequests);
+            BenchmarksEventSource.Measure("http/exceptions", _exceptionRequests);
+            BenchmarksEventSource.Measure("http/rps/mean", (_successRequests + _badStatusRequests) / elapsed);
+
+            if (_successRequests > 0)
+            {
+                LogPercentiledMetric("http/headers", _headersTimes!, TicksToMs);
+                LogPercentiledMetric("http/contentstart", _contentStartTimes!, TicksToMs);
+                LogPercentiledMetric("http/contentend", _contentEndTimes!, TicksToMs);
+            }
+        }
+    }
+
+    private static async Task Get(HttpClient client, CancellationToken token, bool collectMetrics)
+    {
+        var oneByteArray = new byte[1];
+        var drainArray = new byte[81920];
+        var stopwatch = Stopwatch.StartNew();
+        while (!token.IsCancellationRequested)
+        {
+            try
+            {
+                stopwatch.Restart();
+                using var result = await client.GetAsync("/get", HttpCompletionOption.ResponseHeadersRead, token);
+                var headersTime = stopwatch.ElapsedTicks;
+                if (result.IsSuccessStatusCode)
+                {
+                    var content = await result.Content.ReadAsStreamAsync();
+                    var bytesRead = await content.ReadAsync(oneByteArray, token);
+                    var contentStartTime = stopwatch.ElapsedTicks;
+                    while (bytesRead != 0)
+                    {
+                        bytesRead = await content.ReadAsync(drainArray, token);
+                    }
+                    var contentEndTime = stopwatch.ElapsedTicks;
+
+                    if (collectMetrics)
+                    {
+                        _headersTimes!.Add(headersTime);
```

Comment:
I worry about using ConcurrentBag here, and also the interlocked operations below. 

We can avoid any sort of cross-thread data sharing here by collecting stats locally for this Task, and then returning them when the test is complete and aggregating the results from the various Tasks.

Reply:
Also: Aren't these bags going to get big? Like really big?

Reply:
Additionally: I'm not sure how much I care about these metrics. A big part of what these are going to measure is *server* latency. I don't care about that here.

Stats like these are interesting for a load-generating app like WRK to capture and report, because they are measuring the *server* performance. And we don't really care if there's extra overhead in the load generating app because we can just run multiple instances of it (or run it on bigger hardware, etc).

But **we are not trying to build WRK here**. "httpClient" is the perf app. "kestrel" is the equivalent of WRK in this case -- though since this is a client scenario, it's not really a "load generator", more like a "load receiver". So if anything, we should be gathering metrics in the kestrel app, not the httpClient app.

Reply:
I am not sure what we can measure on Kestrel side though.
On the client side, it might be interesting to measure e.g. time between calling SendAsync and sending the request on the wire, but it's not easy to get this info from the outside.
But still I see the value of current metrics in terms of comparison between different HTTP versions, while server latency remains the same.

Do you have some ideas on what can we additionally measure?